### DISCUSSION TITLE
Language (translation) support

### DISCUSSION THREAD

**Discussion by Madis0** [GriefPrevention/GriefPrevention#705] [Ideas]
**As** a player
**I want** to read GriefPrevention messages in my language
**Because** I may not understand English that well.

Therefore, I suggest adding full language support to GriefPrevention, so it would be possible to add languages upstream (by commiting to this repo) and activating on the server by simply setting the locale in config.

Also, it would be nice to [allow user-made subtitles](https://support.google.com/youtube/answer/6054623?hl=en) [on the instruction video](https://www.youtube.com/watch?v=VDsjXB-BaE0).

**Comment by RoboMWM** [comment]
Update: translations can be submitted here: https://github.com/GriefPrevention/GriefPrevention/discussions/categories/translations

---

It's not my instruction video so I can't do that. Might do something for v20 - but you can put anything you want there, so it can be other text instructions (like I do on my server) or your own video.

People are free to submit translations either here or in spigot and I'll add it to the wiki. Nobody seems to have done that though despite a few (before I became the maintainer) who have translated the file before. (I used to allow editing until idiots constantly did random edits to it.)

**Comment by Madis0** [comment]
> People are free to submit translations either here or in spigot and I'll add it to the wiki.

Well, why wiki? Why not have separate strings files for different languages, like many other plugins do? That would make it easier for administrators to apply them, too.

**Comment by RoboMWM** [comment]
> Why not have separate strings files for different languages,

I'm sorry, I don't know what you're asking for. 

I personally modify the messages.yml just to change the wording or to remove messages. I'm not interested in hard-coding language values - the messages can be edited like most other mature plugins.

**Comment by RoboMWM** [comment]
The only reasonable suggestion I can make from that is that you want the plugin to come with a bunch of already-translated files. I'm opposed to that for two reasons:

- clutter in the folder (less simple)
- I don't have a translation team

**Comment by RoboMWM** [comment]
I believe there are translation services out there, like transifex (and some other one that I don't recall). I think I tried setting up one but never really played much with it - idk if it's public so others can see it or not.

**Comment by Madis0** [comment]
> The only reasonable suggestion I can make from that is that you want the plugin to come with a bunch of already-translated files. I'm opposed to that for two reasons:
> 
>     * clutter in the folder (less simple)
> 
>     * I don't have a translation team

Yes, that's what I'm asking for. Clutter in the folder is negligible (you could download the locale on server start) and translation team is not necessary due to crowd-sourcing.

 > I believe there are translation services out there, like transifex (and some other one that I don't recall). I think I tried setting up one but never really played much with it - idk if it's public so others can see it or not.

I assume you mean Transifex and Crowdin - I use both for various projects but prefer the latter. Both also give free plans for FOSS projects. Whether it is public (freely joinable) or private (invitation-based) is up to you, in the settings.
But you don't have to sign up for either, you can continue using GitHub and just merge pull requests as they come.

Here are some popular examples of managing languages:

- [EssentialsX](https://github.com/EssentialsX/Essentials/tree/2.x/Essentials/src) [(now using Crowdin for uploads)](https://crowdin.com/project/essentialsx-official)
- [Jobs](https://github.com/Zrips/Jobs/tree/master/src/main/resources/locale)
- [Modern LWC](https://github.com/Tsuser1/Modern-LWC/tree/master/src/lang)

**Comment by Dum4G** [comment]
Any status on that suggestion? @RoboMWM  In case folder cluttering matters you can follow the opeNLogin approach where
config.yml lets you choose the language and points where to download translations from:

(hash) Language of the plugin.
(hash) For more information: https://github.com/nickuc/OpeNLogin/blob/master/docs/lang.md
languageFile: 'messages_en.yml'

Alternatively, you can store all of translations in a jar file and unpack them if non-english locale is being used like AuthMe does.


Your community will become your translation team and your proofreaders

**Comment by ALiangLiang** [comment]
As @Madis0 said. Translations can contributed by other developers. And also [translated to zh-TW (Traditional Chinese for Taiwan) by me](https://gist.github.com/ALiangLiang/d5770e08a2b0234380cc55be87796119). It's a common way to do the i18n.

If you don't want to change the code. Just linking translated `message.yml` in wiki is an easier way. But you can't manage them directly. And also other developers cannot contribute the translation directly too.

**Reply by RoboMWM** [reply]
> If you don't want to change the code. Just linking translated message.yml in wiki is an easier way. But you can't manage them directly. And also other developers cannot contribute the translation directly too.

The wiki here is locked down yes due to vandals but the new docs repo is open for PRs. I've encouraged other translators to contribute to the wiki (before it was locked down) but it's usually just a singular post that they drop in a spigot thread and then I never hear from them ever again.

**Comment by leadvic** [comment]
Hey there, I'm a bit lost here. I'd love to have this plugin in Spanish, and I'm willing to help translating it, but then where should I place it so that anyone else can translate it easily? And if it is already translated, where should I look for it?

**Reply by Jikoo** [reply]
You'd start a discussion in the [translations category](https://github.com/GriefPrevention/GriefPrevention/discussions/categories/translations) and paste the messages.yml content into a ```` ```yaml ```` code block so people can copy it easily.

**Comment by Ezdrael** [comment]
Why does the "messages.yml" file that I translated get lost in English again after I reload the plugin? How to prevent this?

**Reply by Jikoo** [reply]
Sounds like you made a syntax error in your edits. Use a YAML linter to check for errors before reloading the messages.

**Reply by Ezdrael** [reply]
> Схоже, ви зробили синтаксичну помилку у своїх редагуваннях. Використовуйте YAML-лінтер для перевірки на наявність помилок перед перезавантаженням повідомлень.

It worked. Thank you. I'll translate it now.

**Comment by thiagorigonatti** [comment]
I’m currently implementing [localization support using ResourceBundle and `.properties`](https://github.com/thiagorigonatti/GriefPrevention/tree/add-localization-bundle) files for messages in multiple languages.

So far, translations are available for Brazilian Portuguese (`pt_BR`) and Spanish (`es`).

If you are interested in helping translate the plugin into another language, simply create a file named:

`messages_yourLocale.properties` copied from `messages.properties` and place in `src\main\resources\`

For example:

`messages_de.properties`
`messages_fr.properties`
`messages_zh_TW.properties`

Then submit a pull request with your translation.

Please also follow the repository, as I’m planning to set up a public collaborative translation platform soon.

Thank you!
